cmd/gateway-common.go

```diff
@@ -0,0 +1,325 @@
+func FromMinioClientListBucketV2Result(bucket string, result minio.ListBucketV2Result) ListObjectsV2Info {
+	objects := make([]ObjectInfo, len(result.Contents))
+
+	for i, oi := range result.Contents {
+		objects[i] = FromMinioClientObjectInfo(bucket, oi)
+	}
+
+	prefixes := make([]string, len(result.CommonPrefixes))
+	for i, p := range result.CommonPrefixes {
+		prefixes[i] = p.Prefix
+	}
+
+	return ListObjectsV2Info{
+		IsTruncated: result.IsTruncated,
+		Prefixes:    prefixes,
+		Objects:     objects,
+
+		ContinuationToken:     result.ContinuationToken,
+		NextContinuationToken: result.NextContinuationToken,
+	}
+}
+
+// FromMinioClientListBucketResult converts minio ListBucketResult to ListObjectsInfo
+func FromMinioClientListBucketResult(bucket string, result minio.ListBucketResult) ListObjectsInfo {
+	objects := make([]ObjectInfo, len(result.Contents))
+
+	for i, oi := range result.Contents {
+		objects[i] = FromMinioClientObjectInfo(bucket, oi)
+	}
+
+	prefixes := make([]string, len(result.CommonPrefixes))
+	for i, p := range result.CommonPrefixes {
+		prefixes[i] = p.Prefix
+	}
+
+	return ListObjectsInfo{
+		IsTruncated: result.IsTruncated,
+		NextMarker:  result.NextMarker,
+		Prefixes:    prefixes,
+		Objects:     objects,
+	}
+}
+
+// FromMinioClientListBucketResultToV2Info converts minio ListBucketResult to ListObjectsV2Info
+func FromMinioClientListBucketResultToV2Info(bucket string, result minio.ListBucketResult) ListObjectsV2Info {
+	objects := make([]ObjectInfo, len(result.Contents))
+
+	for i, oi := range result.Contents {
+		objects[i] = FromMinioClientObjectInfo(bucket, oi)
+	}
+
+	prefixes := make([]string, len(result.CommonPrefixes))
+	for i, p := range result.CommonPrefixes {
+		prefixes[i] = p.Prefix
+	}
+
+	return ListObjectsV2Info{
+		IsTruncated:           result.IsTruncated,
+		Prefixes:              prefixes,
+		Objects:               objects,
+		ContinuationToken:     result.Marker,
+		NextContinuationToken: result.NextMarker,
+	}
+}
+
+// ToMinioClientMetadata converts metadata to map[string][]string
+func ToMinioClientMetadata(metadata map[string]string) map[string]string {
+	mm := make(map[string]string)
+	for k, v := range metadata {
+		mm[http.CanonicalHeaderKey(k)] = v
+	}
+	return mm
+}
+
+// ToMinioClientCompletePart converts CompletePart to minio CompletePart
+func ToMinioClientCompletePart(part CompletePart) minio.CompletePart {
+	return minio.CompletePart{
+		ETag:       part.ETag,
+		PartNumber: part.PartNumber,
+	}
+}
+
+// ToMinioClientCompleteParts converts []CompletePart to minio []CompletePart
+func ToMinioClientCompleteParts(parts []CompletePart) []minio.CompletePart {
+	mparts := make([]minio.CompletePart, len(parts))
+	for i, part := range parts {
+		mparts[i] = ToMinioClientCompletePart(part)
+	}
+	return mparts
+}
+
+// ErrorRespToObjectError converts Minio errors to minio object layer errors.
+func ErrorRespToObjectError(err error, params ...string) error {
```

Comment:
Better move it to gateway-s3.go and name it as `S3ToObjectError()` and it can be used in gateway-gcs-anonymous.go too.

Reply:
Sure will do.

Reply:
Oh wait we can't do that again due to cyclical dependency.